### PR TITLE
Refactor completion commit.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
test vsi please

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 1]
```diff
@@ -47,225 +56,124 @@ private void Commit(PresentationItem item, Model model, char? commitChar)
...
-                    // Re-center the view.
-                    using (var textEdit = viewBuffer.CreateEdit(editOptions, reiteratedVersionNumber: null, editTag: null))
-                    {
-                        var caretPoint = this.TextView.GetCaretPoint(this.SubjectBuffer);
-                        if (caretPoint.HasValue)
-                        {
-                            this.TextView.Caret.EnsureVisible();
-                        }
+                        // Move the caret back to where it was prior to committing the item.
+                        //TextView.Caret.MoveTo(new VirtualSnapshotPoint(
```
Comment (dpoeschl) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: Commented out?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 2]
```diff
@@ -47,225 +56,124 @@ private void Commit(PresentationItem item, Model model, char? commitChar)
...
+                    if (characterWasSentIntoBuffer && !completionChange.IncludesCommitCharacter)
                     {
-                        document.Project.Solution.Workspace.ApplyTextChanges(document.Id, changes, cancellationToken);
+                        nextHandler();
                     }
 
-                    formattingTransaction.Complete();
+                    // If the insertion is long enough, the caret will scroll out of the visible area.
+                    // Re-center the view.
+                    // this.TextView.Caret.EnsureVisible();
```
Comment (dpoeschl) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: commented out?

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide This is ready for review.

This change drastically simplifies our completion controller logic.  The idea now is quite simple now and has this behavior:

```c#
// When a character is typed it is *always* sent through to the editor.  This way the
// editor always represents what would have been typed had completion not been involved
// at this point.  That means that if we decide to commit, then undo'ing the commit will
// return you to the code that you would have typed if completion was not up.
//
// The steps we follow for commit are as follows:
//
//      1) send the commit character through to the buffer.
//      2) open a transaction.
//          2a) roll back the text to before the text was sent through
//          2b) commit the item.
//          2c) send the commit character through again.
//          2d) commit the transaction.
//
// 2c is very important.  it makes sure that post our commit all our normal features
// run depending on what got typed.  For example if the commit character was (
// then brace completion may run.  If it was ; then formatting may run.  But, importantly
// this code doesn't need to know anything about that.  Furthermore, becaue that code
// runs within this transaction, then the user can always undo and get to what the code
// would have been if completion was not involved.
```

This means that the completion controller needs no code to handle things like formatting.  It needs no code to try to reverse engineer brace completion.  It doesn't need to munge the the changes passed back by the CompletionService.  It doesn't need to do any special handling of hte caret (unless a completion item wants it to).

**Comment by CyrusNajmabadi** [comment]
I recommend just reading the 'after' for Controller_Commit.cs.  The logic has changed enugh that the diff is not htat useful.

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
test vsi please

File: src/EditorFeatures/Test2/IntelliSense/VisualBasicCompletionCommandHandlerTests.vb [hunk 3]
```diff
@@ -28,7 +28,7 @@ End Class
                 Await state.AssertSelectedCompletionItem("On Error GoTo", description:=String.Format(FeaturesResources.Keyword, "On Error GoTo") + vbCrLf + VBFeaturesResources.OnErrorGotoKeywordToolTip)
                 state.SendTypeChars(" ")
                 Await state.WaitForAsynchronousOperationsAsync()
-                Await state.AssertSelectedCompletionItem("Error GoTo", description:=String.Format(FeaturesResources.Keyword, "Error GoTo") + vbCrLf + VBFeaturesResources.OnErrorGotoKeywordToolTip)
+                Await state.AssertSelectedCompletionItem("On Error GoTo", description:=String.Format(FeaturesResources.Keyword, "On Error GoTo") + vbCrLf + VBFeaturesResources.OnErrorGotoKeywordToolTip)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/VisualBasicCompletionCommandHandlerTests.vb]: Note: this is "regression" that i think is ok.  Namely, because 'on ' matches the text in the completion item, we say "continue filtering and don't commit".  I think this is personally ok because the end experience is the same.  With this approach we also don't have the completion list comitting then popping up immediately with that single item.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingPresentationItem.cs [hunk 4]
```diff
@@ -67,14 +67,7 @@ public DescriptionModifyingPresentationItem(CompletionItem item, CompletionServi
...
             // normally the items that produce multiple changes are not expecting to trigger the behaviors that rely on looking at the text
-            if (change.TextChanges.Length == 1)
-            {
-                return change.TextChanges[0];
-            }
-            else
-            {
-                return new TextChange(item.Span, item.DisplayText);
-            }
+            return change.TextChange;
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingPresentationItem.cs]: No clue what that comment means.

File: src/Features/Core/Portable/Completion/Providers/AbstractMemberInsertingCompletionProvider.cs [hunk 5]
```diff
@@ -57,14 +58,42 @@ public AbstractMemberInsertingCompletionProvider()
             }
 
             var changes = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
-            return CompletionChange.Create(ImmutableArray.CreateRange(changes), newPosition, includesCommitCharacter: true);
+            var change = Collapse(newText, changes.ToList());
+
+            return CompletionChange.Create(change, newPosition, includesCommitCharacter: true);
+        }
+
+        private TextChange Collapse(SourceText newText, List<TextChange> changes)
```
Comment (rchande) [inline src/Features/Core/Portable/Completion/Providers/AbstractMemberInsertingCompletionProvider.cs]: Why is this necessary? Are all providers now expected to return a single TextChange?

**Comment by rchande** [comment]
:+1: 
It's terrifying how much simpler this looks; I wonder if we should have people play with it before merging.

**Comment by CyrusNajmabadi** [comment]
This will address https://github.com/dotnet/roslyn/issues/11858

**Comment by CyrusNajmabadi** [comment]
> Why is this necessary? Are all providers now expected to return a single TextChange?

I have two PRs out currently.  

This pr takes the stance that hte API for completion should be that when you commit you specify a single TextChange you want to perform.  Currently we have 0 providers taht need to perform multiple text changes, and multiple text changes can always be represented a single all-emcompassing text change anyways.

The other PR keeps the api where you can provide many changes, but collapses it itself.  Personally, i'd prefer to use this PR and just tell people that retunring multiple code changes is not desirable.

**Comment by balajikris** [comment]
Looks good to me :+1: Do we need command handler level tests for this to ensure that we do the right things such as insert unformatted text and pass the command down the chain etc. ? That'd ensure unintended changes don't creep in.

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Need senior signoff please.

File: src/Features/Core/Portable/Completion/CompletionChange.cs [hunk 6]
```diff
@@ -44,33 +58,55 @@ private CompletionChange(ImmutableArray<TextChange> textChanges, int? newPositio
...
 
+#pragma warning disable RS0027 // Public API with optional parameter(s) should have the most parameters amongst its public overloads.
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
+        public static CompletionChange Create(
+#pragma warning restore RS0026 // Do not add multiple public overloads with optional parameters
+#pragma warning restore RS0027 // Public API with optional parameter(s) should have the most parameters amongst its public overloads.
+            TextChange textChange,
+            int? newPosition = null,
+            bool includesCommitCharacter = false)
+        {
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionChange.cs]: There is no source or binary compat concern with this overload being added.  The first parameter is non-optional in each overload.  And the first parameter will always distinguish exactly which overlaod is being picked.  The RSXXXX warning are for properly adding overloads where the initial parameters share the same types.

**Comment by CyrusNajmabadi** [comment]
retest windows_vsi_p3_open_prtest  please

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_unit32_prtest  please

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_unit64_prtest please

**Comment by CyrusNajmabadi** [comment]
retest windows_vsi_p3_open_prtest  please

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 7]
```diff
@@ -48,229 +51,124 @@ private void Commit(PresentationItem item, Model model, char? commitChar)
...
                     }
 
                     // adjust the caret position if requested by completion service
-                    if (triggerSnapshotChange.NewPosition != null)
+                    if (completionChange.NewPosition != null)
                     {
-                        var target = new SnapshotPoint(this.SubjectBuffer.CurrentSnapshot, triggerSnapshotChange.NewPosition.Value);
-                        this.TextView.TryMoveCaretToAndEnsureVisible(target);
+                        TextView.Caret.MoveTo(new SnapshotPoint(
+                            this.SubjectBuffer.CurrentSnapshot, completionChange.NewPosition.Value));
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: Is this safe?  Does this position ever need to be mapped forward?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 8]
```diff
@@ -31,17 +29,39 @@ void ICommandHandler<TypeCharCommandArgs>.ExecuteCommand(TypeCharCommandArgs arg
...
+            // runs within this transaction, then the user can always undo and get to what the code
+            // would have been if completion was not involved.
+            //
+            // 2c*: note sending the commit character through to the buffer again can be controlled
+            // by the completion item.  For example, completion items that want to totally handle
+            // what gets output into the buffer can ask for this not to happen.  An example of this
+            // is override completion.  If the user types "override Method(" then we'll want to 
+            // spit out the entire method and *not* also spit out "(" again.
+
+            // In order to support 2a (rolling back), we capture hte state of the buffer before
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: spelling `hte`

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs (hunk 7 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: The contract of 'COmpletionChange' is that you give the NewPosition wrt to the final document after your text change is applied.

File: src/Features/Core/Portable/Completion/CompletionChange.cs [hunk 9]
```diff
@@ -12,8 +14,11 @@ namespace Microsoft.CodeAnalysis.Completion
     public sealed class CompletionChange
     {
         /// <summary>
-        /// The text changes to be applied to the document.
+        /// The text change to be applied to the document.
         /// </summary>
+        public TextChange TextChange { get; }
+
+        [Obsolete("Use TextChange instead")]
```
Comment (Pilchie) [inline src/Features/Core/Portable/Completion/CompletionChange.cs]: You can use `nameof` here.

File: src/Features/Core/Portable/Completion/CompletionChange.cs [hunk 10]
```diff
@@ -23,16 +28,25 @@ public sealed class CompletionChange
...
-        /// True if the changes include the typed character that caused the <see cref="CompletionItem"/> to be committed.
-        /// If false the completion host will determine if and where the commit character is inserted into the document.
+        /// True if the changes include the typed character that caused the <see cref="CompletionItem"/>
+        /// to be committed.  If false the completion host will determine if and where the commit 
+        /// character is inserted into the document.
         /// </summary>
         public bool IncludesCommitCharacter { get; }
 
-        private CompletionChange(ImmutableArray<TextChange> textChanges, int? newPosition, bool includesCommit)
+        private CompletionChange(ImmutableArray<TextChange> textChanges, int? newPosition, bool includesCommitCharacter)
```
Comment (Pilchie) [inline src/Features/Core/Portable/Completion/CompletionChange.cs]: Can't this just be removed and the `Create` overload do the substitution?

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionChange.cs]: sure.

**Comment by Pilchie** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest windows_vsi_p0_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p0_open_prtest/253/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest windows_vsi_p1_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p1_open_prtest/246/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest windows_vsi_p3_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p3_open_prtest/258/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
test vsi please
